lib/Sema/CSSimplify.cpp

```diff
@@ -4960,6 +4960,28 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
                                               Type root, Type value,
                                               ConstraintLocatorBuilder locator,
                                               bool isFavored) {
+  // If this is a subscript with a KeyPath expression, add a constraint that
+  // connects the subscript's root type to the root type of the KeyPath.
+  SmallVector<LocatorPathElt, 4> path;
+  Expr *anchor = locator.getLocatorParts(path);
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
+            for (auto constraint : InactiveConstraints) {
+              if (constraint.getKind() == ConstraintKind::KeyPath &&
+                  constraint.getLocator()->getAnchor() == keyPathExpr) {
+                addConstraint(ConstraintKind::Subtype, root, constraint.getSecondType(), locator);
```

Comment:
This typing rule seems to be duplicating the existing rule in CSGen (after drilling).  Try looking at what `visitUnresolvedMemberExpr` is doing.

Reply:
I'm afraid I don't understand which existing rule you're referring to or what specifically in `visitUnresolvedMemberExpr` you're pointing me to. 😞 I'll need some more help to understand. Sorry!

Reply:
I think we should try our best to do this in CSGen because we are trying to avoid dealing with expressions in the solver at all costs...

Reply:
In CSGen you don't know that this is a KeyPath application—it's just a subscript. Because this is inside the subscript overload, I don't think it can be in CSGen. Or maybe I missing something?

But it should be possible to move it out a level to `ConstraintSystem::resolveOverload` if that's preferable.

Reply:
> In CSGen you don't know that this is a KeyPath application

I think it can be done.  A naive way would be to add an extra "faux root" expression node where you can stash the type variable generated for the `root` expression in the KeyPath.  Then in the post-walk, you can recreate this logic (minus the locator digging) and directly bind the subscript's base to the faux root.  For example (don't actually do this, please clean it up):

```cxx
// ConstraintWalker::walkToExprPost
      if (auto subscript = dyn_cast<SubscriptExpr>(expr)) {
        if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
          if (indexTuple->getNumElements() == 1) {
            if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
              auto type = CG.visit(expr);
              auto &CS = CG.getConstraintSystem();
              auto simplifiedType = CS.simplifyType(type);

              CS.setType(expr, simplifiedType);

              CS.addConstraint(ConstraintKind::Subtype,
                               CS.getType(subscript->getBase()),
                               CS.getType(keyPathExpr->getFauxRoot()),
                               CS.getConstraintLocator(expr));

              return expr;
            }
          }
        }
      }

// ...

// ConstraintWalker::visitKeyPathExpr
case KeyPathExpr::Component::Kind::UnresolvedProperty:
          E->setFauxRoot(/**/);
          CS.setType(E->getFauxRoot(), root);
          LLVM_FALLTHROUGH;
```

Reply:
That assumes that a `[keyPath: <key path>]` expression _is_ a KeyPath application. I was assuming that shouldn't be hardcoded, but it actually is today. i.e., this fails to compile:

```swift
extension String {
	subscript(keyPath: KeyPath<Int, String>) -> String {
		return "🙈"
	}
}

_ = ""[keyPath: \Int.description]
```

Since that's apparently not valid, it should be fine to add this constraint outside of the KeyPath application constraint. (Although that kinda seems like a bug to me. Why shouldn't you be able to overload a subscript that takes a KeyPath?)

Given that…

1. Is there any reason it's preferable to use `walkToExprPost` over `visitSubscriptExpr`? As long as we're going to assume that `[keyPath: <key path>]` is a key path application, we might as well remove the overload and do this directly in `visitSubscriptExpr`?

2. Is it still preferable to create another expression instead of using the locator? Neither seem ideal to me. 😕 That would mean adding a new type like `KeyPathRootExpr`, right?

Reply:
> Is there any reason it's preferable to use walkToExprPost over visitSubscriptExpr

I use it out of convenience.  Anywhere in the walk that has access to the subscript and its children after they have had constraints generated will do.

> As long as we're going to assume that [keyPath: <key path>] is a key path application, we might as well remove the overload and do this directly in visitSubscriptExpr

No such luck in the general case

```swift
extension String {
  subscript<T>(keypath a: T) -> String {
    return "🙈"
  }
}
_ = ""[keypath: 5]
```

> Is it still preferable to create another expression instead of using the locator

Digging in the locators is a hack, adding a fake AST node is a hack.  Neither is desirable.  You shouldn't need a new type for this, the whole goal of the example I gave was that you need to stash the type variable generated for root somewhere that's accessible to the rest of CSGen.

Reply:
The suggestion @CodaFi made it along the lines I was thinking as well, but I'm not sure if we really need to create a faux root through, because even if we still let root allocate new type variable but then connect it to something which can inferred that would still be good enough. Another idea, but I'm not sure how feasible that would be, is to actually make it so `KeyPathExpr`, when created, knows what expression it's attached to, that would be very straightforward to get its type from the constraint system cache and relate it to root in `visitKeyPathExpr`.